File: typescript/packages/core/src/loop/compact.ts
import { assertNever } from "../assert-never";
import type { EventOf } from "../fold/state";
import type { Outcome } from "../hooks/invoke";
import type { EventId, KnownEvent } from "../log";
import { refReader, render } from "../render";
import type { EventDraft } from "../store";
import { type Attempted, attempt } from "./attempt";
import { draft, HOST, type RECOVERY } from "./drafts";
import { decision, defining, recorded, run } from "./hooks";
import { contextPolicy, tokens } from "./policy";
import { restoreDrafts, restoreHooks } from "./restore";
import type { Session } from "./session";
import { cancelRequested, stepEvents } from "./turn";
import type { Halt } from "./types";

// (clear old results) and L2 (summary compaction, then L3 restore), run on a
// threshold by the ladder (ladder.ts), reactively by L4 and L5, or on request (manual.ts).

export type Compaction =
  | { readonly kind: "compacted" }
  | { readonly kind: "failed" }
  /** The summary request's response held a registered secret: the turn has ended. */
  | { readonly kind: "ended" }
  | { readonly kind: "halt"; readonly halt: Halt };

type Reason = "threshold" | "compaction_fallback";
export type FailReason =
  | "still_over_threshold"
  | "empty_summary"
  | "prompt_too_long"
  | "model_error"
  | "artifact_error";

/** Who a compaction's outcome is for: the request it answers, and recovery when it settles it. */
export type Answering = {
  readonly cause?: EventId;
  readonly actor?: typeof HOST | typeof RECOVERY;
};

/** L5's once-per-step guard, shared with L4: this step already compacted or failed to. */
export function reactiveSpent(s: Session): boolean {
  return stepEvents(s.events, s.fold).some(
    (e) => e.type === "compacted" || e.type === "compaction_failed",
  );
}

export async function compact(
  s: Session,
  trigger: "reactive" | "threshold",
): Promise<Compaction> {
  const ctx = contextPolicy(s.fold.policy);
  const cleared = clear(s, ctx.clear_results.keep_recent, "threshold");
  if (cleared !== undefined) return { kind: "halt", halt: cleared };
  const range = compactRange(s);
  if (range === undefined) return failed(s, "still_over_threshold");
  const gated = await beforeCompact(s);
  if (gated !== undefined) return gated;
  const got = await sideRequest(s);
  switch (got.kind) {
    case "halt":
      return got;
    case "response":
      return got.text === ""
        ? failed(s, "empty_summary")
        : summarized(s, range, got.text, trigger);
    case "rejected":
      return failed(
        s,
        got.rejection.reason === "prompt_too_long"
          ? "prompt_too_long"
          : "model_error",
      );
    case "leaked":
      // With a cancel pending the turn is still open: the compaction failed, and the
      // cancellation step closes the turn.
      return got.ended ? { kind: "ended" } : failed(s, "model_error");
    case "broken":
    case "unsupported":
      return failed(s, "model_error");
    case "budget":
      // The attempt recorded the failure with budget_exceeded.
      return { kind: "failed" };
    default:
      return assertNever(got);
  }
}

/**
 * The side request. If it is itself too long, everything clearable is cleared and it is retried
 * once, unless a cancel landed meanwhile: nothing is sent after the barrier, so the rejection
 * stands and the compaction fails.
 */
async function sideRequest(s: Session): Promise<Attempted> {
  const got = await attempt(s, "compaction", 1);
  if (
    got.kind !== "rejected" ||
    got.rejection.reason !== "prompt_too_long" ||
    cancelRequested(s.events) !== undefined
  )
    return got;
  const fallback = clear(s, 0, "compaction_fallback");
  if (fallback !== undefined) return { kind: "halt", halt: fallback };
  return attempt(s, "compaction", 2);
}

/** compaction_failed{summary}, naming the latest side request unless none was made for it. */
export function failed(
  s: Session,
  reason: FailReason,
  answering: Answering = {},
): Compaction {
  const { cause, actor = HOST } = answering;
  const side = s.events.findLast(
    (e) =>
      e.type === "model_request" &&
      e.data.purpose === "compaction" &&
      e.data.cause_event_id === cause,
  );
  const omit =
    side === undefined ||
    reason === "still_over_threshold" ||
    reason === "artifact_error";
  const stopped = s.append(
    draft.compactionFailed(
      {
        stage: "summary",
        reason,
        ...(omit ? {} : { request_event_id: side.event_id }),
        ...(cause === undefined ? {} : { cause_event_id: cause }),
      },
      actor,
    ),
  );
  return stopped === undefined
    ? { kind: "failed" }
    : { kind: "halt", halt: stopped };
}

/**
 * `compacted` over `range`, from the latest side request's summary, with its restore in the
 * same batch; the context hooks follow.
 */
export async function summarized(
  s: Session,
  range: readonly [KnownEvent, KnownEvent],
  text: string,
  trigger: "reactive" | "threshold" | "manual",
  answering: Answering = {},
): Promise<Compaction> {
  const [from, to] = range;
  const { cause, actor = HOST } = answering;
  const side = s.events.findLast(
    (e) => e.type === "model_request" && e.data.purpose === "compaction",
  );
  if (side === undefined) throw new Error("a summary answers a side request");
  const restored = await restoreDrafts(s, from.seq, to.seq);
  const stopped = s.append(
    draft.compacted(
      {
        from_seq: from.seq,
        to_seq: to.seq,
        from_event_id: from.event_id,
        to_event_id: to.event_id,
        summary_ref: s.store(text, "text/plain"),
        summary_request_event_id: side.event_id,
        trigger,
        ...(cause === undefined ? {} : { cause_event_id: cause }),
      },
      actor,
    ),
    ...restored,
  );
  if (stopped !== undefined) return { kind: "halt", halt: stopped };
  const hooked = await restoreHooks(s);
  return hooked === undefined
    ? { kind: "compacted" }
    : { kind: "halt", halt: hooked };
}

/**
 * before_compact gates the side request: a deny or failure is
 * compaction_failed{stage: hook}; a guide is recorded with its text, which the side request's
 * instruction line carries (Render v1). Every extension sees the state from before the first
 * call. For a request, an extension that already decided after it is not asked again.
 */
export async function beforeCompact(
  s: Session,
  request?: EventOf<"compaction_requested">,
): Promise<Compaction | undefined> {
  const window =
    request === undefined ? [] : s.events.filter((e) => e.seq > request.seq);
  const cause =
    request === undefined ? {} : { cause_event_id: request.event_id };
  const state = s.state();
  for (const ext of defining(s, "before_compact")) {
    if (recorded(window, ext.name, "before_compact") !== undefined) continue;
    const out = await run(ext, "before_compact", [state]);
    const made = compactDecision(ext.name, out);
    const denied =
      made.data.decision !== "proceed" && made.data.decision !== "guide";
    const stopped = denied
      ? s.append(
          made,
          draft.compactionFailed({
            stage: "hook",
            reason: "hook_denied",
            ...cause,
          }),
        )
      : s.append(made);
    if (stopped !== undefined) return { kind: "halt", halt: stopped };
    if (denied) return { kind: "failed" };
  }
  return undefined;
}

function compactDecision(
  ext: string,
  out: Outcome<"before_compact">,
): Extract<EventDraft, { type: "hook_decision" }> {
  const hook = "before_compact";
  if (out.kind === "failed")
    return decision(ext, hook, "failed", {}, out.reason);
  const v = out.value;
  switch (v.decision) {
    case "proceed":
      return decision(ext, hook, "proceed");
    case "guide":
      return decision(ext, hook, "guide", {}, v.text);
    case "deny":
      return decision(ext, hook, "deny", {}, v.reason);
    default:
      return assertNever(v);
  }
}

/**
 * L1: every result rendered in an earlier turn request, except the `keep` newest and those
 * already cleared, is replaced by the fixed placeholder.
 */
export function clear(
  s: Session,
  keep: number,
  reason: Reason,
): Halt | undefined {
  const events = s.events;
  const ctx = contextPolicy(s.fold.policy);
  const lastRequest = events.findLast(
    (e) => e.type === "model_request" && e.data.purpose !== "compaction",
  );
  const done = new Set(
    events.flatMap((e) =>
      e.type === "context_edited"
        ? e.data.edits.filter((x) => x.action === "clear").map((x) => x.call_id)
        : [],
    ),
  );
  const results = events.flatMap((e) =>
    e.type === "tool_result" && e.seq < (lastRequest?.seq ?? 0)
      ? [e.data.call_id]
      : [],
  );
  const excluded = new Set(ctx.clear_results.exclude_tools);
  const names = new Map(
    events.flatMap((e) =>
      e.type === "tool_call" ? [[e.data.call_id, e.data.name] as const] : [],
    ),
  );
  const ids = results
    .slice(0, Math.max(0, results.length - keep))
    .filter((id) => !done.has(id) && !excluded.has(names.get(id) ?? ""));
  if (ids.length === 0) return undefined;
  return s.append(
    draft.contextEdited({
      reason,
      edits: ids.map((call_id) => ({ call_id, action: "clear" })),
    }),
  );
}

/**
 * L2 range: from the first input after thread_started to the event before the shortest
 * step-boundary tail whose rendered estimate reaches keep_tail.
 */
function compactRange(
  s: Session,
): readonly [KnownEvent, KnownEvent] | undefined {
  const events = s.events;
  const from = events.find((e) => e.type === "user_input");
  if (from === undefined) return undefined;
  const keep = tokens(
    contextPolicy(s.fold.policy).compact.keep_tail,
    windowTokens(s),
  );
  const read = refReader(s.artifacts);
  const whole = render(events, read);
  if (!whole.ok) return undefined;
  for (const to of events.toReversed()) {
    if (to.seq < from.seq) return undefined;
    if (s.fold.boundaries[to.seq] !== true) continue;
    const head = render(
      events.filter((e) => e.seq <= to.seq),
      read,
    );
    if (!head.ok) return undefined;
    // ponytail: re-renders per candidate, O(n^2); keep per-line byte counts if sessions get long.
    const tail = whole.value.bytes.length - head.value.bytes.length;
    if (Math.ceil(tail / 4) >= keep) return [from, to];
  }
  return undefined;
}

/** W: the epoch model's context window minus reserve_tokens. */
export function windowTokens(s: Session): number {
  const model = s.fold.model;
  const listed = s.fold.policy?.models?.find(
    (m) => m.provider === model?.provider && m.name === model.name,
  );
  const adapter = model === undefined ? undefined : s.config.models(model);
  const window =
    listed?.context_window ?? adapter?.info.limits.context_window ?? 0;
  return window - contextPolicy(s.fold.policy).reserve_tokens;
}

File: python/src/threads/loop/compact.py
"""Reactive compaction (with L1 and L2): once per step after the provider says the
prompt is too long. Every layer is an appended event; nothing edits history.

The summarizer call is an ordinary recorded attempt (`model_request{purpose: compaction}`), so
it replays from the log like any request and counts toward cost. The proactive layers that
call into it are loop/ladder.py; a requested compaction is loop/manual.py.
"""

import math
from collections.abc import Sequence
from dataclasses import dataclass
from typing import TYPE_CHECKING, Literal

from threads.hooks.runner import COMPACT, decision_draft, run_one
from threads.hooks.types import wire_name
from threads.log import (
    CallId,
    CompactionRequestedEvent,
    ContextEditedEvent,
    Event,
    EventId,
    HookDecisionEvent,
    ModelAttemptAbandonedEvent,
    ModelRequestEvent,
    ModelResponseEvent,
    ModelResponseRecoveredEvent,
    TextPart,
    ThreadStartedEvent,
    ToolCallEvent,
    ToolResultEvent,
)
from threads.loop import attempt, defaults
from threads.loop.drafts import ActorKind, draft
from threads.loop.gates import said, verdict
from threads.loop.restore import restore_drafts, restore_hooks
from threads.loop.runtime import Failed, Halt, Runtime, lost
from threads.result import Err, Ok

if TYPE_CHECKING:
    from pydantic import JsonValue

type Trigger = Literal["threshold", "reactive", "manual"]


@dataclass(frozen=True, slots=True)
class Answering:
    """Who a compaction's outcome is for: its trigger, the compaction_requested it answers, and
    recovery when recovery settles it."""

    trigger: Trigger = "threshold"
    cause: EventId | None = None
    actor: ActorKind = "host"

    def tag(self, data: "dict[str, JsonValue]") -> "dict[str, JsonValue]":
        return data if self.cause is None else {**data, "cause_event_id": self.cause}


async def reactive(rt: Runtime) -> Halt | None:
    """L1, then L2 with trigger reactive. Ends appending `compacted` or `compaction_failed`; the
    loop then re-sends the turn request, or ends the turn context_exhausted."""
    ctx = defaults.context(rt.fold)
    halt = await clear(rt, ctx.clear_results.keep_recent, "threshold")
    return halt or await summarize(rt, "reactive")


async def summarize(rt: Runtime, trigger: Trigger) -> Halt | None:
    """L2: before_compact, the recorded side request, `compacted`, then L3 restore."""
    bounds = await _range(rt)
    if bounds is None:
        return await failed(rt, "still_over_threshold", None, Answering(trigger))
    denied = await before_compact(rt)
    if denied is not False:
        return denied
    return await _side_request(rt, bounds, trigger)


async def before_compact(
    rt: Runtime, request: CompactionRequestedEvent | None = None
) -> Halt | Literal[False] | None:
    """A deny (or a failure) is compaction_failed{hook, hook_denied}; a guide's text is its
    decision's reason, which the side request's instruction line carries (Render v1). Each
    decision is appended before the next extension runs, and every extension sees the state
    from before the first call. For a request, an extension that already decided after it is not
    asked again. False: go on to the side request."""
    decided = _decided(rt.events, request)
    state = rt.writer.state()
    for bound in rt.hooks.defining("before_compact"):
        if bound.extension in decided:
            continue
        ran = await run_one(bound, "before_compact", COMPACT, state)
        decision = verdict(ran)
        drafts = [
            decision_draft(
                "before_compact", ran, decision, said(ran, "text") or said(ran, "reason")
            )
        ]
        denied = decision == "deny"
        if denied:
            cause = Answering(cause=None if request is None else request.event_id)
            drafts.append(
                draft("compaction_failed", cause.tag({"stage": "hook", "reason": "hook_denied"}))
            )
        done = await rt.append(*drafts)
        if isinstance(done, Err):
            return lost(done.error)
        if denied:
            return None
    return False


def _decided(events: Sequence[Event], request: CompactionRequestedEvent | None) -> set[str]:
    """Extensions with a before_compact decision recorded after the request."""
    if request is None:
        return set()
    hook = wire_name("before_compact")
    return {
        e.data.extension
        for e in events
        if isinstance(e, HookDecisionEvent) and e.seq > request.seq and e.data.hook == hook
    }


async def _side_request(rt: Runtime, bounds: tuple[Event, Event], trigger: Trigger) -> Halt | None:
    """The side request; if it is itself too long, clear every clearable result and retry it
    once."""
    for side_attempt in (1, 2):
        sent = await attempt.request(rt, side_attempt, "compaction")
        if sent is None or isinstance(sent, Failed):
            return sent
        outcome = rt.events[-1]
        if isinstance(outcome, ModelResponseEvent):
            text = response_text(outcome)
            if not text:
                return await failed(rt, "empty_summary", sent, Answering(trigger))
            return await summarized(rt, bounds, sent, text, Answering(trigger))
        too_long = isinstance(outcome, ModelAttemptAbandonedEvent) and (
            outcome.data.reason == "prompt_too_long"
        )
        if not too_long or side_attempt == 2:  # noqa: PLR2004 - the fallback retries once
            reason = "prompt_too_long" if too_long else "model_error"
            return await failed(rt, reason, sent, Answering(trigger))
        halt = await clear(rt, 0, "compaction_fallback")
        if halt is not None:
            return halt
    return None


def _clearable(events: Sequence[Event], keep_recent: int, exclude: Sequence[str]) -> list[CallId]:
    """Results rendered in an earlier turn request, beyond the newest `keep_recent`, not excluded
    and not already cleared."""
    names = {e.data.call_id: e.data.name for e in events if isinstance(e, ToolCallEvent)}
    cleared = {
        edit.call_id
        for e in events
        if isinstance(e, ContextEditedEvent)
        for edit in e.data.edits
        if edit.action == "clear"
    }
    last_request = max(
        (
            e.seq
            for e in events
            if isinstance(e, ModelRequestEvent) and e.data.purpose != "compaction"
        ),
        default=0,
    )
    rendered = [
        e.data.call_id for e in events if isinstance(e, ToolResultEvent) and e.seq < last_request
    ]
    older = rendered[: max(0, len(rendered) - keep_recent)]
    return [c for c in older if c not in cleared and names.get(c) not in exclude]


def clearable(rt: Runtime) -> bool:
    """L1 has something to clear at the configured keep_recent."""
    ctx = defaults.context(rt.fold).clear_results
    return bool(_clearable(rt.events, ctx.keep_recent, ctx.exclude_tools))


async def clear(rt: Runtime, keep_recent: int, reason: str) -> Halt | None:
    """L1: one context_edited clearing every clearable result."""
    exclude = defaults.context(rt.fold).clear_results.exclude_tools
    calls = _clearable(rt.events, keep_recent, exclude)
    if not calls:
        return None
    edits: list[JsonValue] = [{"call_id": c, "action": "clear"} for c in calls]
    done = await rt.append(draft("context_edited", {"reason": reason, "edits": edits}))
    return lost(done.error) if isinstance(done, Err) else None


async def _range(rt: Runtime) -> tuple[Event, Event] | None:
    """From the first event after thread_started to the end of the shortest tail that starts at a
    step boundary and still holds `keep_tail` tokens (by estimate: bytes / 4)."""
    events = rt.events
    started = next(i for i, e in enumerate(events) if isinstance(e, ThreadStartedEvent))
    first = started + 1
    keep = defaults.tokens(defaults.context(rt.fold).compact.keep_tail, rt.fold)
    whole = await rt.store.render(events)
    if isinstance(whole, Err) or first >= len(events):
        return None
    for end in range(len(events) - 1, first - 1, -1):
        # A response's own tool calls follow it: ending there would split the pair.
        splits = end + 1 < len(events) and isinstance(events[end + 1], ToolCallEvent)
        if events[end].seq not in rt.fold.boundaries or splits:
            continue
        head = await rt.store.render(events[: end + 1])
        if not isinstance(head, Ok):
            return None
        if math.ceil((len(whole.value.body) - len(head.value.body)) / 4) >= keep:
            return events[first], events[end]
    return None


def response_text(response: ModelResponseEvent | ModelResponseRecoveredEvent) -> str:
    return "".join(p.text for p in response.data.content if isinstance(p, TextPart))


async def summarized(
    rt: Runtime, bounds: tuple[Event, Event], request_id: EventId, text: str, answering: Answering
) -> Halt | None:
    """`compacted` over `bounds` with its restore in the same batch; the context hooks follow."""
    raw = text.encode("utf-8")
    sha = await rt.store.put_artifact(raw)
    first, last = bounds
    data: dict[str, JsonValue] = {
        "from_seq": first.seq,
        "to_seq": last.seq,
        "from_event_id": first.event_id,
        "to_event_id": last.event_id,
        "summary_ref": {"sha256": sha, "bytes": len(raw), "media_type": "text/plain"},
        "summary_request_event_id": request_id,
        "trigger": answering.trigger,
    }
    dropped = [e for e in rt.events if first.seq <= e.seq <= last.seq]
    restored = await restore_drafts(rt, dropped)
    done = await rt.append(draft("compacted", answering.tag(data), answering.actor), *restored)
    if isinstance(done, Err):
        return lost(done.error)
    return await restore_hooks(rt)


async def failed(
    rt: Runtime, reason: str, request_id: EventId | None, answering: Answering
) -> Halt | None:
    data: dict[str, JsonValue] = {"stage": "summary", "reason": reason}
    if request_id is not None:
        data["request_event_id"] = request_id
    done = await rt.append(draft("compaction_failed", answering.tag(data), answering.actor))
    return lost(done.error) if isinstance(done, Err) else None
